File: myFinance_improved_V2/lib/features/cash_ending/data/models/currency_model.dart
// lib/features/cash_ending/data/models/currency_model.dart

import '../../domain/entities/currency.dart';
import 'denomination_model.dart';

/// Data Transfer Object for Currency
/// Handles JSON serialization/deserialization
class CurrencyModel {
  final String currencyId;
  final String currencyCode;
  final String currencyName;
  final String symbol;
  final List<DenominationModel> denominations;

  const CurrencyModel({
    required this.currencyId,
    required this.currencyCode,
    required this.currencyName,
    required this.symbol,
    this.denominations = const [],
  });

  /// Create from JSON (from database)
  factory CurrencyModel.fromJson(Map<String, dynamic> json) {
    return CurrencyModel(
      currencyId: json['currency_id']?.toString() ?? '',
      currencyCode: json['currency_code']?.toString() ?? '',
      currencyName: json['currency_name']?.toString() ?? '',
      symbol: json['symbol']?.toString() ?? '',
      denominations: (json['denominations'] as List<dynamic>?)
              ?.map((d) => DenominationModel.fromJson(d as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  /// Convert to JSON (to database)
  Map<String, dynamic> toJson() {
    return {
      'currency_id': currencyId,
      'currency_code': currencyCode,
      'currency_name': currencyName,
      'symbol': symbol,
      'denominations': denominations.map((d) => d.toJson()).toList(),
    };
  }

  /// Convert to Domain Entity
  Currency toEntity() {
    return Currency(
      currencyId: currencyId,
      currencyCode: currencyCode,
      currencyName: currencyName,
      symbol: symbol,
      denominations: denominations.map((d) => d.toEntity()).toList(),
    );
  }

  /// Create from Domain Entity
  factory CurrencyModel.fromEntity(Currency entity) {
    return CurrencyModel(
      currencyId: entity.currencyId,
      currencyCode: entity.currencyCode,
      currencyName: entity.currencyName,
      symbol: entity.symbol,
      denominations: entity.denominations
          .map((d) => DenominationModel.fromEntity(d))
          .toList(),
    );
  }
}

File: website/src/features/cash-ending/data/models/CurrencyModel.ts
/**
 * CurrencyModel
 * Handles JSON transformation between API and Domain Entity
 */

import { CurrencyEntity, type CurrencyWithExchangeRate } from '../../domain/entities/Currency';

export class CurrencyModel {
  /**
   * Convert API JSON response to Domain Entity
   */
  static fromJson(json: CurrencyWithExchangeRate): CurrencyEntity {
    return new CurrencyEntity(
      json.currency_id,
      json.company_currency_id,
      json.currency_code,
      json.currency_name,
      json.symbol,
      json.flag_emoji,
      json.is_base_currency,
      json.exchange_rate_to_base,
      json.denominations,
      json.created_at
    );
  }

  /**
   * Convert Domain Entity to API JSON format
   */
  static toJson(entity: CurrencyEntity): CurrencyWithExchangeRate {
    return {
      currency_id: entity.currencyId,
      company_currency_id: entity.companyCurrencyId,
      currency_code: entity.currencyCode,
      currency_name: entity.currencyName,
      symbol: entity.symbol,
      flag_emoji: entity.flagEmoji,
      is_base_currency: entity.isBaseCurrency,
      exchange_rate_to_base: entity.exchangeRateToBase,
      denominations: entity.denominations,
      created_at: entity.createdAt,
    };
  }
}
